src/ray/object_manager/pull_manager.h

```diff
@@ -0,0 +1,264 @@
+  /// object size and so can't always fill out this field in the constructor.
+  ///
+  /// \param num_chunks The number of chunks.
+  /// \return Void.
+  void InitializeChunksIfNecessary(int64_t num_chunks);
+
+  /// Check if this PullInfo object can safely be cleaned up because the object
+  /// is not required, and we are not receiving the object from anywhere, and
+  /// there are no chunks in the process of being read for this object.
+  ///
+  /// \return True if this object can be cleaned up and false otherwise.
+  bool LifetimeEnded();
+
+  /// Restart the timer. This happens when we successfully read a chunk of the
+  /// object or when we become aware of new object locations (because we want to
+  /// give more time for the transfer to finish before retrying). It also can
+  /// happen when the timer expires.
+  void RestartTimer(boost::asio::io_service &main_service);
+
+  /// True if we must pull this object. False if we are simply receiving the
+  /// object but do not need to pull the object (meaning we do not have to
+  /// guarantee that the object appears locally).
+  bool required;
+  /// Our most recent estimate of which object managers have the object.
+  std::unordered_set<ClientID> clients_with_object;
+  /// The IDs of the remote object managers that we have already requested
+  /// the object from. If we cancel a request, then we will remove that
+  /// client from this set.
+  std::unordered_set<ClientID> clients_requested_from;
+  /// If this object manager is currently receiving the object from a remote
+  /// object manager, this will be the client ID of the remote object
+  /// manager. Otherwise, it will be nil.
+  ClientID client_receiving_from;
+  /// The total number of chunks that the object is divided into. If this is
+  /// -1, then the number is not known.
+  int64_t total_num_chunks;
+  /// The chunk IDs that have successfully been received.
+  std::unordered_set<int> received_chunk_ids;
+  /// The chunks that have not yet been received. This is the complement of
+  /// the values in received_chunk_ids.
+  std::unordered_set<int> remaining_chunk_ids;
+  /// The number of object chunks that we are in the process of reading.
+  int64_t num_in_progress_chunk_ids;
+  /// This timer is used for two purposes: 1) If we are receiving the
+  /// object from some remote object manager but one of the reads fails,
+  /// then this timer will be used to issue new requests. If we have requested
+  /// the object but haven't received any chunks in a while, then we consider
+  /// the sending to have failed.
+  boost::asio::deadline_timer retry_timer;
+  /// The callback to call when the timer expires.
+  std::function<void()> timer_callback;
+};
+
+using ObjectRequestManagementCallback = std::function<void(
+    const std::vector<ray::ClientID> &clients_to_request,
+    const std::vector<ray::ClientID> &clients_to_cancel, bool abort_object)>;
+
+/// This class is responsible for ensuring that objects that this object
```

Comment:
This header was quite difficult for me to parse. One suggestion for making this clearer is to start with a high-level description of the invariants (e.g., for every object, this object manager should only be receiving it from 0 or 1 other object managers). Then follow with a state machine description for an individual object, to show how we uphold the invariants (e.g., how we go from `NOT_RECEIVING` to `RECEIVING`, but cannot go from `RECEIVING` to `RECEIVING` from another object manager).

Reply:
Here's my TLDR:

PullManager tracks the state of pull requests (i.e., requests to get this object on this node, or speculative push requests of an object to this node). Objects can only be pulled from one other node at a time. PullManager keeps a timer per pull request that is used to handle timed out remote object managers (???). Non-speculative pull requests will be retried indefinitely, while push requests are aborted if the pusher goes away.

This is just my impression based on reading the header file here, I haven't looked at the implementation yet.